### DISCUSSION TITLE
Difference between the user_code_launcher and run_launcher?

### DISCUSSION THREAD

**Discussion by skeller88** [dagster-io/dagster#18355] [Q&A] [answered]
The difference is that [run_launcher](https://docs.dagster.io/deployment/run-launcher#run-launchers) is set in your local `dagster.yaml` config or your Dagster cloud config, and [user_code_launcher](https://docs.dagster.io/dagster-cloud/managing-deployments/setting-environment-variables-dagster-cloud-agents#amazon-ecs-agents-1) is specific to the internals of how the agent launches workers? 

On that note, what's the reason for the terminology of "user code" instead of something like a "worker" or "executor"? A lot of people I'm showing [this diagram too](https://docs.dagster.io/dagster-cloud/deployment/hybrid) are finding the "user code" boxes a bit confusing.

**Comment by gibsondan** [comment, marked as answer]
The reason that it's called a user code launcher rather than a run launcher is that it's responsible for launching both runs and code servers, which are responsible for serving requests for sensor ticks and other low-latency access to Dagster code. In Dagster Cloud you just configure a user code launcher, you don't directly configure a run launcher.

Thanks for the feedback about the terminology being confusing, I'll pass that on to the rest of the team.

**Reply by marcusthoreson-UG** [reply]
So then if I am, say, setting up a code location in Dagster Cloud (Hybrid deployment) with an on-prem Docker agent, I have to use the same container image for the code server and the launched runs, correct? With a Dagster OSS deployment I could configure a DockerRunLauncher to use separate images for the code server and run containers; is there any way to achieve the same configuration in Dagster Cloud?

**Reply by gibsondan** [reply]
Dagster generally assumes that the code in the code server that defines your jobs is the same as the code in the run container - unexpected behavior can happen if they get out of sync. If you want your job to run code in a different container image, you could consider something like Dagster Pipes (https://docs.dagster.io/guides/dagster-pipes)

(to more directly answer your question, that's not currently possible with a Docker agent - but I'm not sure i'd recommend it in Dagster OSS either, even if it's possible)

**Comment by skeller88** [comment]
Is a code server a run worker / executor
<https://docs.dagster.io/deployment/overview#job-execution-flow>? Dagster
Cloud describes that you configure an agent
<https://docs.dagster.io/dagster-cloud/deployment/agents#dagster-cloud-hybrid-agents>,
which is described as a run launcher (which is actually a user code
launcher according to this thread :)).

**Reply by gibsondan** [reply]
No, a code server is not a run worker or an executor.

**Reply by gibsondan** [reply]
It’s the “Code location server” mentioned on that same page: https://docs.dagster.io/deployment/overview
